### PR TITLE
Engine concept loading + better STI model support

### PR DESCRIPTION
**Pull request by rosskevin**
- Allow concepts to be loaded from engines
- Re-sort load order so that all models are first (better support for STI models).  Note this is a 99% solution, for outliers you may still need to be explicit and use `require_dependency`

This closes #13, #14, #30.

cc: @apotonick

### CODE REVIEW

**Comment by sveredyuk** [comment]
I need all this stuff... My app is built totally on engines, so I meet a lot of troubles with moving it to Trailblazer.

File: lib/trailblazer/rails/railtie.rb [hunk 1]
```diff
@@ -3,8 +3,37 @@
 
 module Trailblazer
   class Railtie < ::Rails::Railtie
+    def self.load_root(root)
+
+      # Loader has no concept for base dir, must chdir so that it can find files local to the root
+      Dir.chdir(root) do
+        # For STI models, we need to sort all models before concepts, gather files first from loader
+        files = []
+        Loader.new.({debug: false, insert: [ModelFile, before: Loader::AddConceptFiles]}) { |file|
```
Comment (apotonick) [inline lib/trailblazer/rails/railtie.rb]: Shouldn't we have a new function `ModelFiles` instead of `ModelFile` that simply gathers all model files in one step, instead of loading them per concept? This would reduce logic, wouldn't it?

Comment (sveredyuk) [inline lib/trailblazer/rails/railtie.rb]: I check this "workaround" in the project and seem it doesn't help. Models are loading later than concepts.... 😢

Comment (sveredyuk) [inline lib/trailblazer/rails/railtie.rb]: RailsEngines deliver so much pain... But we need to support it I think engine-orientation + trailblazer it's a one and only way to built normal rails app.

Comment (rosskevin) [inline lib/trailblazer/rails/railtie.rb]: @apotonick I could not see a way to plug this in as-is to the Loader, but I agree, that would make more sense.  

We could load _everything_ in the model directory first, not just what is hinted from the concept (which would further solve some edge cases), then just load the concepts.  

I was just trying to work with what was there.  Perhaps you can take a shot at what you would like and I'll test it out?

Comment (aiomaster) [inline lib/trailblazer/rails/railtie.rb]: @apotonick For me this feature would be usefull, too. Even if it is not 100% perfect, the solution from @rosskevin works for me. What if you accept this PR, release a new version and optimize later (or let others optimize)? (Maybe I don't fully understand the problem on loading the models per concept. Is it just a performance hit or are there other traps?)

Comment (apotonick) [inline lib/trailblazer/rails/railtie.rb]: @rosskevin I think "blindly" loading the models first would resolve most problems, but we should offer that as an alternative loading strategy. I will take a stab at your code and rearrange it a bit. 

@aiomaster I hear you, but that would result in even more work for me since some semantics will definitely change when we "optimize" the code.

**Comment by apotonick** [comment]
@rosskevin @sveredyuk @aiomaster It would be fantastic if you could test against trailblazer-loader/master and trailblazer-rails/master. :v:

**Comment by sveredyuk** [comment]
@rosskevin @aiomaster @apotonick i found strange thing in development env. Every time when i change the code inside some engine concepts folder i need restart web-server.

**Comment by apotonick** [comment]
@sveredyuk With the new code? I didn't even know that Rails reloads engine code automatically... I thought you always have to restart?

**Comment by sveredyuk** [comment]
@apotonick Not really. If your engine located locally in Gemfile through `path: 'your_engine/path` than it reload it code on change.

**Comment by apotonick** [comment]
Ok, then we might have to find out how to register that for reloads. Sounds very familiar.

However, can you confirm that the new code by @rosskevin and myself now actually loads engine concepts? :beers:

**Comment by sveredyuk** [comment]
@apotonick Seems that no...
```
gem 'trailblazer-rails',
  git: 'git@github.com:trailblazer/trailblazer-rails.git',
  branch: :master

gem 'trailblazer-loader',
  git: 'git@github.com:trailblazer/trailblazer-loader.git',
  branch: :master
```
```
lib/ruby/gems/2.1.0/bundler/gems/trailblazer-rails-c1573f3d61b9/lib/trailblazer/rails/railtie.rb:71:in `block in <class:Railtie>': wrong number of arguments (1 for 2) (ArgumentError)
```

**Comment by rosskevin** [comment]
Sorry for my lateness @apotonick.   What seems to be missed here (as I commented in the commit) is that you are only loading the models in _this_ project, not the ones from dependencies as well.

Here is the relevant code from my commit:

```
      # Iterate over all engines and load concepts
      engines = ::Rails::Engine.subclasses.map(&:instance)
      engines.each do |engine|
        load_root(engine.root)
      end
````

**Comment by rosskevin** [comment]
One problem, I'll PR th glob change and anything else I need shortly.

**Comment by rosskevin** [comment]
@see #33

**Comment by rosskevin** [comment]
The only real caveat to this is that operations that inherit from other operations (and override methods such as `process`) can get overwritten due to load order.  It's a strange edge case occurrence, and even though the hierarchy shouldn't be violated, for some reason, a more-base-class required later appears to overwrite the implementation of the subclass's method.  Explicit use of `require_dependency` solves it.

I don't have a better solution to this, other than to suggest adding a warning in the readme.  On the outside, it seems like trb isn't working, but diving in shows the load order problem.

**Comment by apotonick** [comment]
But overally it works, right?

**Comment by rosskevin** [comment]
Yes, works fine now with 33 merged. Thank you
